Add BiMultiMap to `rule_structs.rs` and use it in RuleManager Move

suryatho opened the pull request:
- RwLocks into RuleManager for more granular control of blocking
- Implemented borrow for Topic struct so we strings can lend their ownership to topic when all we really need is &Topic

## Changes

See #475 

## Notes

I didn't really test this yet.

## Test Cases

- Add rules for client and remove all rules for said client, then try to add rules for that client.

_If you did any manual testing (e.g., with Postman), put screenshots of the http request and before and after of the db_

## To Do

_Any remaining things that need to get done_

- [ ] Test properly

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `package-lock.json` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #475

suryatho committed:
Add BiMultiMap to `rule_structs.rs` and use it in RuleManager Move
RwLocks into RuleManager for more granular control of blocking Implement
borrow for Topic struct so we strings can lend their ownership to topic
when all we really need is &Topic

suryatho commented:
I pushed my initial changes for the RuleManager, but I wasn't too sure how I should go about testing it (because I don't know how to create rules). It would be nice if you could tell me how you would like me to test it. It builds tho

suryatho committed:
Update BiMultiMap to return optional of newly empty values when removing
key. Used to remove useless rules from topic and rules maps in
RuleManager.

bracyw commented:
> I pushed my initial changes for the RuleManager, but I wasn't too sure how I should go about testing it (because I don't know how to create rules). It would be nice if you could tell me how you would like me to test it. It builds tho

Definitely will, I'll plan to spend some time going over in Thursday meeting. For now if you want to spend some time outside of some proper guidance on this, look into using postman to test adding/deleting rules (via http) and subscribing to rules via the socket output. You should be able to use a fake client_id.

suryatho committed:
Improve BiMultiMap api to remove get_mut methods and also have clearer
method signatures on remove methods. Fix logical error in delete_rule
and delete_client, where entire topic was removed from index when rule
is no longer subscribed to.

suryatho committed:
Use tokio::join! wherever possible for better concurrency

suryatho committed:
Use generic internal static methods for BiMultiMap

suryatho committed:
Use branching concurrency inside handle_msg

suryatho commented:
oh sounds cool, I will try to figure out how to make my own rules and send them using Postman. It doesn't seem too hard we even have a gwee for it.

One more question about the logic for the RuleManager:
I've written it up such that whenever you call `delete_rule(client, rule)` (for deleting a rule from a client), if the client no longer has any more rules then it is removed from `subscriptions` entirely, and if the rule is no longer serving anybody it also gets kicked out the data structure as a whole. Similarly when we `delete_client(client)` (for removing a client as a key from subscriptions), if any of the rules it was subscribed to no longer serve any clients, those rules also get kicked from the data structure. 

I forgot if you said something else which is why I am asking this question, but this was the way I figured it should be implemented.

So basically the question is: Whenever a rule no longer serves any clients, should it be deleted from all parts of the data structure? (Or should we keep it in the `topic_index` or `rules` maps) And if a client no longer is asking for any rules after a delete call should that client be removed from `subscriptions`? (It should right)

bracyw commented:
> oh sounds cool, I will try to figure out how to make my own rules and send them using Postman. It doesn't seem too hard we even have a gwee for it.
> 
> One more question about the logic for the RuleManager: I've written it up such that whenever you call `delete_rule(client, rule)` (for deleting a rule from a client), if the client no longer has any more rules then it is removed from `subscriptions` entirely, and if the rule is no longer serving anybody it also gets kicked out the data structure as a whole. Similarly when we `delete_client(client)` (for removing a client as a key from subscriptions), if any of the rules it was subscribed to no longer serve any clients, those rules also get kicked from the data structure.
> 
> I forgot if you said something else which is why I am asking this question, but this was the way I figured it should be implemented.
> 
> So basically the question is: Whenever a rule no longer serves any clients, should it be deleted from all parts of the data structure? (Or should we keep it in the `topic_index` or `rules` maps) And if a client no longer is asking for any rules after a delete call should that client be removed from `subscriptions`? (It should right)

It's looking good! 

(for commits can you try and model them after some of the closed PR's like single sentence with `#XXX` (ticket number in front)

For the first part I don't think we have a GUI, Argos 2 (a flutter a app Jack made) does have a GUI I believe.

For the deletion how bout we have a force delete on empty list argument  (boolean true if you want to auto delete, false if not) (so you can control whether or not you want to force delete only when you want to).

suryatho commented:
Oh I thought I had put #475 in each of the commit messages, I can do that then.

I meant like that postman has a gui for the making of these requests. (it does right ?)

Where would the force delete be? Is it on the `BiMultiMap` api such that it can leave keys with no values. Or do you mean on the api for RuleManager where we can leave rules that no longer serve any clients in the `topic_index` and `rules` maps.

bracyw commented:
> I meant like that postman has a gui for the making of these requests. (it does right ?)

oh yeah postman does, lmao (one day you will have to try out Argos 2, it's in flutter)


just in the BiMultiMap and for now the RuleManager won't delete, unless specifically asked (for a rule). So that way unsubscribing won't delete a rule auto. And for now we won't support for deleting a rule through unsubscribing client side.

suryatho commented:
> just in the BiMultiMap and for now the RuleManager won't delete, unless specifically asked (for a rule). So that way unsubscribing won't delete a rule auto. And for now we won't support for deleting a rule through unsubscribing client side.

So using this logic: there will never be mappings in the `subscriptions` bimap where a client is mapped to an empty set of rules and vice versa, but we will allow rules to exist in `topic_index` and `rules` without serving any clients (not in `subscriptions`). So basically from the client side we can always add rules but never delete them from the data structure entirely (we remove can from subscriptions). 

that makes sense!

I may have read what you meant wrong. But if I didn't, the solution is probably pretty easy, and that we never actually delete from the rule_cache and I can get rid of that code (Or should I keep it as part of the API for future?).

suryatho committed:
Make sure rule manager can't delete rule by client unsubscribing to rule
manager #475

suryatho commented:
Oh when I was committing before the # in the message was being interpreted as a comment in the message.

bracyw commented:
> > just in the BiMultiMap and for now the RuleManager won't delete, unless specifically asked (for a rule). So that way unsubscribing won't delete a rule auto. And for now we won't support for deleting a rule through unsubscribing client side.
> 
> So using this logic: there will never be mappings in the `subscriptions` bimap where a client is mapped to an empty set of rules and vice versa, but we will allow rules to exist in `topic_index` and `rules` without serving any clients (not in `subscriptions`). So basically from the client side we can always add rules but never delete them from the data structure entirely (we remove can from subscriptions).
> 
> that makes sense!
> 
> I may have read what you meant wrong. But if I didn't, the solution is probably pretty easy, and that we never actually delete from the rule_cache and I can get rid of that code (Or should I keep it as part of the API for future?). I would just have to support adding rules that are already part of the cache more efficiently instead (upsert/don't insert if we already have a rule "cached").

Exactly what I was thinking for the unsub / sub functionality! As far as the rule_cach, deleting should still happen just not from a unsubscribing, only from an explicit rule deletion. (I guess what I was saying about sub/unsub for now doesn't matter because all the endpoints are oriented at a client adding or deleting rules (not unsubscribing which will come later).

suryatho commented:
I don't think we have an actual **_delete_** rule method on the `RuleManager` API in the sense that there is a method called `delete_rule`, but it was used previously (and still is) as a method to delete a rule from a client, or in other words, unsubscribe the client from a rule (unrelated but we do have `delete_client`). I don't think it would be fitting to delete the rule from the rules and topic cache here since this method seems like its only meant to remove a relationship of one client to one of it's rules. Do you suggest I write a real delete rule method that actually removes the rule data entirely from the RuleManager.

Maybe I am missing something here about the final architecture of the rule system, but wouldn't it be easiest to implement if the action of a client unsubscribing from a rule could lead to the deletion of that rule if no more clients are subscribed to that rule, and if the client disconnects or is deleted the rules it was subscribed to could be deleted if they aren't subscribed to by other clients.

bracyw commented:
> I don't think we have an actual **_delete_** rule method on the `RuleManager` API in the sense that there is a method called `delete_rule`, but it was used previously (and still is) as a method to delete a rule from a client, or in other words, unsubscribe the client from a rule (unrelated but we do have `delete_client`). I don't think it would be fitting to delete the rule from the rules and topic cache here since this method seems like its only meant to remove a relationship of one client to one of it's rules. Do you suggest I write a real delete rule method that actually removes the rule data entirely from the RuleManager.
> 
> 
> 
> Maybe I am missing something here about the final architecture of the rule system, but wouldn't it be easiest to implement if the action of a client unsubscribing from a rule could lead to the deletion of that rule if no more clients are subscribed to that rule, and if the client disconnects or is deleted the rules it was subscribed to could be deleted if they aren't subscribed to by other clients. 

The idea I was going for is that a client can unsubscribe from a rule and still go back to it later. If they really don't want it then they can delete. But transferring the logic is all that is required for this PR (so whatever is there logic wise we should keep (in terms of function))

suryatho committed:
Add test file for testing functionality of bimultimap and rule manager
in rule_structs.rs #475

suryatho commented:
I added a test file and wrote in some unit tests (it passes all tests), which was almost all AI but there were some important tests in there, notably: concurrent adding of rules, and testing of the handle_msg method in cases where there are mutiple clients as well. Let me know how much I should trim down the test file. I also had to make some changes in `rule_structs.rs` regarding the access levels of certain members.

bracyw commented on scylla-server/tests/rule_structs_test.rs:
```diff
@@ -0,0 +1,789 @@
+        values: vec![15.0], // a = 15.0 > 10, should trigger
+        timestamp: Utc::now(),
+    };
+
+    // First trigger might not fire due to debounce logic
+    let _ = rule_manager.handle_msg(&client_data).await;
+
```
This should test that there is no notification returned (and maybe just make debounce and wait time a little longer to ensure test isn't flaky (i.e. can fail sometimes depending on speed of computer))

bracyw replied:
IK test under get's to this but probably should check here also for clarity

bracyw commented on scylla-server/tests/rule_structs_test.rs:
```diff
@@ -0,0 +1,789 @@
+        handle.await.unwrap()?;
+    }
+
+    assert_eq!(rule_manager.get_all_rules().await.len(), 10);
```
should check that all were added, plus length.

bracyw commented on scylla-server/tests/rule_structs_test.rs:
```diff
@@ -0,0 +1,789 @@
+}
+
+#[tokio::test]
+async fn test_rule_manager_concurrent_operations() -> Result<(), RuleManagerError> {
```
can you make a couple more of these tests

bracyw reviewed (commented):
Looks good, added a few comments on tests. No need to delete any of them, however maybe put bi-multi-map tests either at end of file or in a different file.

suryatho committed:
Added more concurrent tests and more asserts #475

suryatho commented:
let me know what I should remove from the test files they're lowk very long.

suryatho committed:
Remove lots of redundant testst #475

suryatho committed:
Remove redundant tests from bimultimap tests #475

suryatho committed:
Make handle_msg more performant ? #475

suryatho commented:
I can't tell if my latest commit is like bad code or not but I was thinking you could do the stuff in handle_msg as an unordered set of tasks so I wanted to implement that. But that is also relying on the fact that we can read from rule managers from multiple tasks so the code looks weird? (or its fine idrk).

bracyw commented:
> I can't tell if my latest commit is like bad code or not but I was thinking you could do the stuff in handle_msg as an unordered set of tasks so I wanted to implement that. But that is also relying on the fact that we can read from rule managers from multiple tasks so the code looks weird? (or its fine idrk).

I would say the idea isn't bad (spawning multiple tasks to increase concurrent processing), but for now I think we should keep things relatively simple, because I don't think processing time will be a problem, rather possible bugs with Rwlocks which will be more problematic with multiple task spawning within the handle message function. After you switch back I'll be giving this a final review after Tuesday (I got a few things at work and stuff that I need to focus on until then.)

After this we can finalize our endpoints and frontend design, super exciting!

suryatho committed:
Undo changes to handle_msg #475

bracyw commented on scylla-server/src/rule_structs.rs:
```diff
@@ -10,17 +10,173 @@ use rustc_hash::FxHashSet;
+    right_to_left: FxHashMap<R, FxHashSet<L>>,
+}
+
+#[allow(dead_code)]
```
shouldn't silence warnings, either comment out / delete unused functions

bracyw commented on scylla-server/src/rule_structs.rs:
```diff
@@ -169,57 +340,86 @@ impl Default for RuleManager {
+        let rule_ids = match self.topic_index.read().await.get(&data.name) {
+            Some(rule_ids) => rule_ids.clone(), // Clone so we can drop resource
+            None => {
+                trace!("Could not find rule in topic -> rule index: {}", data.name);
```
can remove this trace

suryatho replied:
i think they are supposed to be warn! 's. Thats what its been like in other spots where we return errors

bracyw commented on scylla-server/src/rule_structs.rs:
```diff
@@ -169,57 +340,86 @@ impl Default for RuleManager {
+            let (triggered_result, clients_result) = {
+                // Future for if rule was triggered
+                let triggered_future = async {
+                    if let Some(rule) = self.rules.write().await.get_mut(&rule_id) {
```
We will end up having to switch to a better data structure for rules, especially if there are performance issues–like you said before–but for now I think we can just delete the futures as they have to wait for the write lock anyway and instead just check before this if there are any clients attached to the rule to avoid the rules write lock unless there are subscribers. 

TLDR: remove futures and just wait for the locks, move clients check above rules check and abort check of rule if no one is subbed.

bracyw commented on scylla-server/src/rule_structs.rs:
```diff
@@ -169,57 +340,86 @@ impl Default for RuleManager {
+                tokio::pin!(clients_future);
+
+                // Check which operation finished first
+                tokio::select! {
```
this wouldn't be needed if we delete those futures.

bracyw commented on scylla-server/src/rule_structs.rs:
```diff
@@ -228,105 +428,86 @@ impl RuleManager {
-                let mut map = FxHashMap::default();
-                map.insert(rule.topic.clone(), vec![rule]);
-                self.clients_map.insert(client, map);
+            async {
```
this can also just be sequential for now. No need for async.

bracyw commented on scylla-server/tests/rule_structs_test.rs:
```diff
@@ -0,0 +1,544 @@
+
+    if let Some(notifications) = result {
+        // Both rules should trigger since 15.0 > 10 and 15.0 > 5
+        assert!(notifications.len() >= 1);
```
this is a little flaky because they should either both always trigger or not, so maybe an assert eq

bracyw commented on scylla-server/tests/rule_structs_test.rs:
```diff
@@ -0,0 +1,544 @@
+
+        if let Ok(Some(notifications)) = result {
+            // Count how many rules should trigger for this value
+            let expected_triggers = num_rules - (value as usize / 10).min(num_rules);
```
This I believe is wrong should be this:

      let expected_triggers = ((value as usize / 10).min(num_rules - 1)) + 1;

bracyw reviewed (changes requested):
couple small changes. lmk if you have questions

suryatho committed:
Remove some parallel bits in favor of sequential #475

suryatho committed:
Formatting #475

bracyw reviewed (approved):
LGTM, nice work! Feel free to merge.
